kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/expressions/InExpressionEvaluator.java

```diff
@@ -0,0 +1,353 @@
+                    + "Non-literal expression found at position %d: %s. "
+                    + "Only constant values are currently supported in IN lists.",
+                i + 1, child.getClass().getSimpleName()));
+      }
+    }
+  }
+
+  private static void validateTypeCompatibility(
+      In in, DataType valueDataType, List<DataType> inListDataTypes) {
+    for (int i = 0; i < inListDataTypes.size(); i++) {
+      DataType listElementType = inListDataTypes.get(i);
+      if (!areTypesCompatible(valueDataType, listElementType)) {
+        throw unsupportedExpressionException(
+            in,
+            String.format(
+                "IN expression requires all elements to have compatible types. "
+                    + "Value type: %s, incompatible element type at position %d: %s. "
+                    + "Consider casting the incompatible element to a compatible type.",
+                valueDataType, i + 1, listElementType));
+      }
+    }
+  }
+
+  /** Validates that collation is only used with string types and the collation is UTF8Binary. */
+  private static void validateCollation(
+      In in,
+      DataType valueDataType,
+      List<Expression> inListExpressions,
+      List<DataType> inListDataTypes) {
+    in.getCollationIdentifier()
+        .ifPresent(
+            collationIdentifier -> {
+              checkIsUTF8BinaryCollation(in, collationIdentifier);
+              validateStringTypesForCollation(
+                  in, valueDataType, inListExpressions, inListDataTypes);
+            });
+  }
+
+  private static void validateStringTypesForCollation(
+      In in,
+      DataType valueDataType,
+      List<Expression> inListExpressions,
+      List<DataType> inListDataTypes) {
+    checkIsStringType(
+        valueDataType, in, "'IN' with collation expects STRING type for the value expression");
+    for (int i = 0; i < inListDataTypes.size(); i++) {
+      if (!isNullLiteral(inListExpressions.get(i))) {
+        checkIsStringType(
+            inListDataTypes.get(i),
+            in,
+            "'IN' with collation expects STRING type for all list elements");
+      }
+    }
+  }
+
+  private static void checkIsUTF8BinaryCollation(In in, CollationIdentifier collationIdentifier) {
+    if (!"SPARK.UTF8_BINARY".equals(collationIdentifier.toString())) {
+      throw unsupportedExpressionException(
+          in,
+          String.format(
+              "Unsupported collation: \"%s\". "
+                  + "Default Engine supports just \"SPARK.UTF8_BINARY\" collation.",
+              collationIdentifier));
+    }
+  }
+
+  private static void checkIsStringType(DataType dataType, In in, String message) {
+    if (!(dataType instanceof StringType)) {
+      throw unsupportedExpressionException(in, message);
+    }
+  }
+
+  @VisibleForTesting
+  static boolean areTypesCompatible(DataType type1, DataType type2) {
+    return type1.equivalent(type2) || belongToSameTypeGroup(type1, type2);
+  }
```

Comment:
Is this all that is required? It seems like we use `valueType` as targetType, shouldn't we require all children are the same or narrower?

Please correct me if I'm missing something in this code

Reply:
Also, do we really support this across TimestampType and TimestampNtz? I'm surprised